server/conn.go

```diff
@@ -447,11 +468,13 @@ func (cc *clientConn) Run() {
 		}
 
 		cc.alloc.Reset()
-		data, err := cc.readPacket()
+		data, err := cc.readPacketWithTimeout()
 		if err != nil {
 			if terror.ErrorNotEqual(err, io.EOF) {
 				errStack := errors.ErrorStack(err)
-				if !strings.Contains(errStack, "use of closed network connection") {
+				if strings.Contains(errStack, "i/o timeout") {
+					log.Infof("con:%d read packet timeout, close this connection.", cc.connectionID)
+				} else if !strings.Contains(errStack, "use of closed network connection") {
```

Comment:
It's better use this way to detect timeout error

```
		start := time.Now()
		data, err := cc.readPacketWithTimeout()
		idleTime := time.Now().Sub(start)
		if err != nil {
			if terror.ErrorNotEqual(err, io.EOF) {
				if netErr, isNetErr :=  errors.Cause(err).(net.Error); isNetErr && netErr.Timeout() {
					log.Infof("con:%d read packet timeout, close this connection, idle: %v, timeout: %v", cc.connectionID, idleTime, cc.getSessionVarsWaitTimeout())
				} else {
					errStack := errors.ErrorStack(err)
					if !strings.Contains(errStack, "use of closed network connection") {
						log.Errorf("con:%d read packet error, close this connection %s",
							cc.connectionID, errStack)
					}
				}
			}
			return
		}
```

detect close maybe be improve in furture too - - https://github.com/golang/go/issues/4373